checkers/badRegexp_checker.go

```diff
@@ -66,11 +68,64 @@ func (c *badRegexpChecker) checkPattern(pat string) {
 		return
 	}
 
-	c.exprFlags = c.exprFlags[:0]
+	c.flagStates = c.flagStates[:0]
+	c.goodAnchors = c.goodAnchors[:0]
+
+	// In Go all flags (modifiers) are set to false by default,
+	// so we start from the empty flag set.
+	c.flagStates = append(c.flagStates, regexpFlagState{})
 
+	c.markGoodCarets(re.Expr)
+	c.markGoodDollars(re.Expr)
 	c.walk(re.Expr)
 }
 
+func (c *badRegexpChecker) markGoodCarets(e syntax.Expr) {
+	canSkip := func(e syntax.Expr) bool {
```

Comment:
can be a method, isn't it?

Reply:
I failed to find a good name for it since it's only used inside this method context.. :(

Reply:
ok, lgtm